source/common/http/utility.cc

```diff
@@ -29,8 +29,145 @@
+    ASSERT(it.identifier().value() <= std::numeric_limits<uint16_t>::max());
+    switch (it.identifier().value()) {
+    case NGHTTP2_SETTINGS_ENABLE_PUSH:
+      if (it.value().value() == 1) {
+        throw EnvoyException("server push is not supported by Envoy and can not be enabled via a "
+                             "SETTINGS parameter.");
+      }
+      break;
+    case NGHTTP2_SETTINGS_ENABLE_CONNECT_PROTOCOL:
+      // This parameter can not be validated due to the use of proto3 bool, which doesn't allow
+      // checking for presence.
+      // TODO: This can be fixed via a breaking API change of the type to
+      // `google.protobuf.BoolValue`.
+      break;
+    case NGHTTP2_SETTINGS_HEADER_TABLE_SIZE:
+      if (options.has_hpack_table_size()) {
+        duplicate_parameters.push_back("hpack_table_size");
+      }
+      break;
+    case NGHTTP2_SETTINGS_MAX_CONCURRENT_STREAMS:
+      if (options.has_max_concurrent_streams()) {
+        duplicate_parameters.push_back("max_concurrent_streams");
+      }
+      break;
+    case NGHTTP2_SETTINGS_INITIAL_WINDOW_SIZE:
+      if (options.has_initial_stream_window_size()) {
+        duplicate_parameters.push_back("initial_stream_window_size");
+      }
+      break;
+    default:
+      // Ignore unknown parameters.
+      break;
+    }
+  }
+
+  if (!duplicate_parameters.empty()) {
+    throw EnvoyException(fmt::format(
+        "the {{{}}} HTTP/2 SETTINGS parameter(s) can not be configured through both named and "
+        "custom "
+        "parameters",
+        absl::StrJoin(duplicate_parameters, ",", [](std::string* out, const std::string& str) {
+          absl::StrAppend(out, str);
+        })));
+  }
+}
+
+} // namespace
+
+const uint32_t OptionsLimits::MIN_HPACK_TABLE_SIZE;
+const uint32_t OptionsLimits::DEFAULT_HPACK_TABLE_SIZE;
+const uint32_t OptionsLimits::MAX_HPACK_TABLE_SIZE;
+const uint32_t OptionsLimits::MIN_MAX_CONCURRENT_STREAMS;
+const uint32_t OptionsLimits::DEFAULT_MAX_CONCURRENT_STREAMS;
+const uint32_t OptionsLimits::MAX_MAX_CONCURRENT_STREAMS;
+const uint32_t OptionsLimits::MIN_INITIAL_STREAM_WINDOW_SIZE;
+const uint32_t OptionsLimits::DEFAULT_INITIAL_STREAM_WINDOW_SIZE;
+const uint32_t OptionsLimits::MAX_INITIAL_STREAM_WINDOW_SIZE;
+const uint32_t OptionsLimits::MIN_INITIAL_CONNECTION_WINDOW_SIZE;
+const uint32_t OptionsLimits::DEFAULT_INITIAL_CONNECTION_WINDOW_SIZE;
+const uint32_t OptionsLimits::MAX_INITIAL_CONNECTION_WINDOW_SIZE;
+const uint32_t OptionsLimits::DEFAULT_MAX_OUTBOUND_FRAMES;
+const uint32_t OptionsLimits::DEFAULT_MAX_OUTBOUND_CONTROL_FRAMES;
+const uint32_t OptionsLimits::DEFAULT_MAX_CONSECUTIVE_INBOUND_FRAMES_WITH_EMPTY_PAYLOAD;
+const uint32_t OptionsLimits::DEFAULT_MAX_INBOUND_PRIORITY_FRAMES_PER_STREAM;
+const uint32_t OptionsLimits::DEFAULT_MAX_INBOUND_WINDOW_UPDATE_FRAMES_PER_DATA_FRAME_SENT;
+
+envoy::config::core::v3::Http2ProtocolOptions
+initializeAndValidateOptions(const envoy::config::core::v3::Http2ProtocolOptions& options) {
```

Comment:
Should we validate the `custom_settings_parameters` as well? This may be better than getting resets from the server.

Reply:
It is not possible to generically validate the custom parameters as they can be use case specific.